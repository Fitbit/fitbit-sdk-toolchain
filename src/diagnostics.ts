import { relative } from 'path';

import chalk from 'chalk';
import dateformat from 'dateformat';
import indentString from 'indent-string';

interface LineAndCharacter {
  line: number;
  character: number;
}

export interface DiagnosticMessage {
  messageText: string;
  category: DiagnosticCategory;
}

export interface Diagnostic {
  messageText: string | DiagnosticMessage[];
  category: DiagnosticCategory;
  target?: DiagnosticTarget;
  file?: {
    path: string;
    position?:
      | {
          start: { line: number };
        }
      | {
          start: LineAndCharacter;
          end?: LineAndCharacter;
        };
  };
}

export enum DiagnosticCategory {
  Warning = 0,
  Error = 1,
  Suggestion = 2,
  Message = 3,
}

export enum DiagnosticTarget {
  App = 'app',
  Companion = 'companion',
  Settings = 'settings',
}

export type DiagnosticHandler = (diagnostic: Diagnostic) => void;

const consoleColors = {
  [DiagnosticCategory.Warning]: chalk.keyword('orange'),
  [DiagnosticCategory.Error]: chalk.red,
  [DiagnosticCategory.Message]: chalk,
  [DiagnosticCategory.Suggestion]: chalk,
};

const categoryNames = {
  [DiagnosticCategory.Warning]: 'warn',
  [DiagnosticCategory.Error]: 'error',
  [DiagnosticCategory.Message]: 'info',
  [DiagnosticCategory.Suggestion]: 'suggestion',
};

export const logDiagnosticToConsole: DiagnosticHandler = (diagnostic) => {
  const { category, messageText, target } = diagnostic;

  let filePosition = '';
  if (diagnostic.file) {
    const { path, position } = diagnostic.file;
    filePosition += ` ${relative(process.cwd(), path)}`;
    if (position) {
      filePosition += `:${position.start.line + 1}`;
      if ('character' in position.start) {
        filePosition += `,${position.start.character + 1}`;
      }
    }
  }

  const now = new Date();
  const timestamp = dateformat(now, 'HH:MM:ss');

  const colorizer = consoleColors[category];
  const categoryName = categoryNames[category];
  const targetString = target ? target : 'build';
  const message =
    typeof messageText === 'string' ? messageText : messageText[0].messageText;

  console.log(
    colorizer(
      `[${timestamp}][${categoryName}][${targetString}]${filePosition} ${message}`,
    ),
  );

  if (typeof messageText !== 'string') {
    for (let i = 1; i < messageText.length; i += 1) {
      console.log(
        consoleColors[messageText[i].category](
          indentString(messageText[i].messageText, 4),
        ),
      );
    }
  }
};
