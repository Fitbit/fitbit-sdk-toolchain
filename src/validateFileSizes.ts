import stream from 'stream';

import PluginError from 'plugin-error';
import Vinyl from 'vinyl';

import { DiagnosticCategory, DiagnosticHandler } from './diagnostics';
import byteStringFormatter from './util/byteStringFormatter';

const PLUGIN_NAME = 'validateFileSizes';

export default function validateFileSizes({
  maxSizes,
  onDiagnostic,
}: {
  maxSizes: Record<string, number | undefined>;
  onDiagnostic: DiagnosticHandler;
}) {
  return new stream.Transform({
    objectMode: true,
    transform(this: stream.Transform, file: Vinyl, _, cb) {
      const expectedMaxSize = maxSizes[file.relative];
      if (file.isNull() || expectedMaxSize === undefined) {
        return cb(undefined, file);
      }

      if (!file.isBuffer()) {
        // Error if file is not a buffer, may in the future support file streams
        return cb(
          new PluginError(
            PLUGIN_NAME,
            `File is not a buffer: ${file.relative}`,
            {
              fileName: file.relative,
            },
          ),
        );
      }

      const actualSize = file.contents.byteLength;
      if (actualSize > expectedMaxSize) {
        onDiagnostic({
          category: DiagnosticCategory.Error,
          messageText: `${
            file.relative
          } is larger than maximum allowed size. File size was ${byteStringFormatter(
            actualSize,
            4,
          )}, maximum allowed is ${byteStringFormatter(expectedMaxSize, 4)}.`,
        });
      }

      return cb(undefined, file);
    },
  });
}
