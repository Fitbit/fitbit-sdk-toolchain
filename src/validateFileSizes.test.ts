import { Readable } from 'stream';

import Vinyl from 'vinyl';

import { DiagnosticHandler } from './diagnostics';
import validateFileSizes from './validateFileSizes';

const TEST_FILENAME = 'test.bin';

const mockDiagnosticHandler = jest.fn();
let validateFileSizesParams: {
  maxSizes: Record<string, number>;
  onDiagnostic: DiagnosticHandler;
};

beforeEach(() => {
  validateFileSizesParams = {
    maxSizes: {
      [TEST_FILENAME]: 3 * 1024 * 1024,
    },
    onDiagnostic: mockDiagnosticHandler,
  };
});

describe('in streaming mode', () => {
  it('throws an error', (done) => {
    expect.assertions(1);
    const fileStream = new Readable({
      read() {
        done();
      },
    });
    // Node 8 will never fire the read above if there isn't
    // at least 1 byte pushed here
    fileStream.push(Buffer.alloc(1));

    const validator = validateFileSizes(validateFileSizesParams);

    validator
      .on('data', () => {
        return;
      })
      .on('error', (error) => expect(error).toMatchSnapshot())
      .end(
        new Vinyl({
          path: 'test.bin',
          contents: fileStream,
        }),
      );
  });
});

describe('in buffered mode', () => {
  it('passes through all files untouched', (done) => {
    const files = [
      new Vinyl({
        path: 'a.js',
        contents: Buffer.of(),
      }),
      new Vinyl({
        path: 'directory',
        stat: {
          isDirectory: () => true,
        } as any,
      }),
    ];

    const handleData = jest.fn();

    const validator = validateFileSizes(validateFileSizesParams);

    validator
      .on('error', done.fail)
      .on('data', handleData)
      .on('end', () => {
        expect(handleData).toHaveBeenCalledTimes(files.length);
        files.forEach((value, index) =>
          expect(handleData).toHaveBeenNthCalledWith(index + 1, value),
        );
        done();
      });

    files.forEach((file) => validator.write(file));
    validator.end();
  });

  // FIXME: Jest typings are broken and can't deal with the done callback parameter properly
  // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34617
  it.each<any>([
    ['one byte over the allowed size', 3 * 1024 * 1024 + 1],
    ['much larger than the allowed size', 10 * 1024 * 1024],
  ])('logs a warning if a file is %s', (_, fileSize, done) => {
    expect.assertions(1);

    const validator = validateFileSizes(validateFileSizesParams);

    validator
      .on('error', done.fail)
      .on('data', jest.fn())
      .on('end', () => {
        expect(mockDiagnosticHandler.mock.calls[0]).toMatchSnapshot();
        done();
      });

    validator.write(
      new Vinyl({
        path: TEST_FILENAME,
        contents: Buffer.alloc(fileSize),
      }),
    );
    validator.end();
  });
});
