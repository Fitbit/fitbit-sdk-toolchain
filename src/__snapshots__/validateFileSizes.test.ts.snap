// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in buffered mode logs a warning if a file is much larger than the allowed size 1`] = `
Array [
  Object {
    "category": 1,
    "messageText": "test.bin is larger than maximum allowed size. File size was 10 MB, maximum allowed is 3 MB.",
  },
]
`;

exports[`in buffered mode logs a warning if a file is one byte over the allowed size 1`] = `
Array [
  Object {
    "category": 1,
    "messageText": "test.bin is larger than maximum allowed size. File size was 3 MB, maximum allowed is 3 MB.",
  },
]
`;

exports[`in streaming mode throws an error 1`] = `
PluginError {
  "fileName": "test.bin",
  "message": "File is not a buffer: test.bin",
  "name": "Error",
  "plugin": "validateFileSizes",
}
`;
